src/Psl/DateTime/Timestamp.php

```diff
@@ -0,0 +1,266 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Psl\DateTime;
+
+use DateTime as NativeDateTime;
+use DateTimeZone as NativeDateTimeZone;
+use IntlDateFormatter;
+use Psl\Locale;
+use Psl\Math;
+
+use function hrtime;
+use function strtotime;
+use function time;
+
+/**
+ * Represents a precise point in time, with seconds and nanoseconds since the Unix epoch.
+ */
+final class Timestamp implements TemporalInterface
+{
+    use \Psl\DateTime\TemporalConvenienceMethodsTrait;
+
+    /**
+     * @var null|array{seconds: int, nanoseconds: int}
+     *
+     * @internal
+     */
+    private static ?array $offset = null;
+
+    /**
+     * @param int $seconds
+     * @param int<0, 999999999> $nanoseconds
+     */
+    private function __construct(
+        private readonly int $seconds,
+        private readonly int $nanoseconds,
+    ) {
+    }
+
+    /**
+     * Creates a new instance for the current moment.
+     *
+     * Returns the current time with precision up to nanoseconds, adjusted to maintain accuracy.
+     *
+     * @return self The current timestamp.
+     *
+     * @pure
+     */
+    public static function now(): self
+    {
+        $hrTime = hrtime();
```

Comment:
is hrtime the right tool to use here over microtime?
As far as I understand, it should solely be used for performance metrics (since it's counted from an arbitrary point in time) rather than for timestamps.

You are playing with the offset, which I don't fully understand.

Reply:
Hrtime allows us to calculate the nanoseconds between every call to `now()`, so when you retrieve two points in time, you will be able to calculate the difference between them down to the nanosecond.

Psl is not the only lib that does this, see: https://github.com/symfony/clock/blob/7.0/MonotonicClock.php

<details>
<summary>How hrtime/offset works</summary>
Yes, hrtime is from an arbitry point of time, but its precious!

to explain what we do here, lets say time() returns `100`, while hrtime() returns `[2, 4000]`

after 2 seconds, and 1000 ns have passed, `time()` would return `102`, while `hrtime()` would return `[4, 5000]`.

so the first time `now()` is called, we get the hrtime:

```
$hrtime = hrtime(); // [2, 4000]
```

check if we have previously set the different between `time()` and `hrtime()`, if not, calculate the different
```
$time = time(); // 100
$different = [
  'seconds' => $time - $hrtime[0], // 100 - 2 = 98
  'nanoseconds' => $hrtime[1],
];
```

then, we calcualte the time in both seconds an nanoseconds, like this:

```
$seconds = $hrtime[0] + $different['seconds']; // 2 + 98 = 100
$nanoseconds = $hrtime[1] + $different['nanoseconds]; // 4000 + 4000 = 8000;

```

so, the first call to `now()` results in `[100, 8000]`

after `1000` ns have passed, if we call `now()` again, what happens is:

```
$hrtime = hrtime(); // [2, 5000]
```

we then add offset:

```
$seconds = $hrtime[0] + $different['seconds']; // 2 + 98 = 100
$nanoseconds = $hrtime[1] + $different['nanoseconds]; // 5000 + 4000 = 9000;
```

so the result we get is `[100, 9000]`, much more accurate than regular `time()`, allowing you to calcualte the difference between 2 points of time down to the nanosecond!

but, there is more!

lets say 10 seconds, and 999999999 nanoseconds have passed!


```
$hrtime = hrtime(); // [12, 999999999]
```

we then add offset:

```
$seconds = $hrtime[0] + $different['seconds']; // 12 + 98 = 110
$nanoseconds = $hrtime[1] + $different['nanoseconds]; // 999999999 + 4000 = 1000004000;
```

this later gets normalized to `[111, 4000]`
</details>

Reply:
Thanks for the explanation. Makes sense indeed!
Do you need the `hrtime() === false` check like symfony has here as well and is there a fallback? (for example no nanoseconds will be used - or up to microseconds)?

Reply:
Yep, i have added a check, and i also split this function into two, `now()` and `monotonic()`, `now()` is enough for almost all cases, while `monotonic()` is useful for things like performance measurment.

note: `now()` uses `microtime()`, so while it is not precious to the nanoseconds, it is precious to the millisecond, making it a better alternative to `time()`.